From 37789b0cd5ee262678507827b7548c038e2a52e1 Mon Sep 17 00:00:00 2001
From: Ivan Pavlotsky <ivan.pavlotsky@lge.com>
Date: Fri, 2 May 2025 20:25:45 +0300
Subject: [PATCH] [webos][upstreamable] 137.0.7104.0~1: V8: GCC: Explicitely
 cast NEON intrinsics in simd.cc

:Release Notes:
Explicitely cast NEON intrinsics in v8's simd.cc to fix webOS build
issue after rolling to 137.0.7104.0~1

:Detailed Notes:
In the upstream v8 CL
https://chromium-review.googlesource.com/c/v8/v8/+/6397652 which is
rolled in Chromium in the upstream CL https://crrev.com/c/6422880 used
improper NEON intrinsics types conversions in simd.cc so GCC complains
about it when building webOS:
../v8/src/objects/simd.cc:932:36: error: cannot convert 'uint8x16_t' to 'uint16x8_t'
...
../v8/src/objects/simd.cc:1002:20: error: cannot convert 'uint64x2_t' to 'uint8x16_t' in initialization

Provided patch via DEPS where used excplicit NEON intrinsics types
conversions to fix the issue.

:Issues Addressed:
[NEVA-10281] Upgrade neva-main-rolling up to 7108 BUILD
---
 src/objects/simd.cc | 31 +++++++++++++++++++++++++++++++
 1 file changed, 31 insertions(+)

diff --git a/src/objects/simd.cc b/src/objects/simd.cc
index 3174ae27b88..222d8e0e02c 100644
--- a/src/objects/simd.cc
+++ b/src/objects/simd.cc
@@ -929,7 +929,13 @@ bool Uint8ArrayFromHexWithNeon(base::Vector<T>& input_vector,
       uint8x16_t second_part_first_batch =
           vld1q_u8(reinterpret_cast<const uint8_t*>(&input_vector[i + 8]));
       first_batch =
+// NOTE(neva): Fix for compilation with GCC on ARM.
+#if !defined(clang)
+          vmovn_high_u16(vmovn_u16(vreinterpretq_u16_u8(first_batch)),
+                         vreinterpretq_u16_u8(second_part_first_batch));
+#else   // !defined(clang)
           vmovn_high_u16(vmovn_u16(first_batch), second_part_first_batch);
+#endif  // defined(clang)
     }
 
     // Load second batch of 16 hex characters into a Neon register
@@ -942,7 +948,13 @@ bool Uint8ArrayFromHexWithNeon(base::Vector<T>& input_vector,
       uint8x16_t second_part_second_batch =
           vld1q_u8(reinterpret_cast<const uint8_t*>(&input_vector[i + 24]));
       second_batch =
+// NOTE(neva): Fix for compilation with GCC on ARM.
+#if !defined(clang)
+          vmovn_high_u16(vmovn_u16(vreinterpretq_u16_u8(second_batch)),
+                         vreinterpretq_u16_u8(second_part_second_batch));
+#else   // !defined(clang)
           vmovn_high_u16(vmovn_u16(second_batch), second_part_second_batch);
+#endif  // defined(clang)
     }
 
     // low nibbles are values with even indexes in fist_batch.
@@ -960,15 +972,28 @@ bool Uint8ArrayFromHexWithNeon(base::Vector<T>& input_vector,
     // Append low nibbles of first batch and second batch and remove 0x00s.
     // (0x36, 0x66, 0x46, 0x32, 0x31, 0x32, 0x31, 0x32, 0x36, 0x66, 0x66, 0x32,
     // 0x31, 0x32, 0x31, 0x66)
+// NOTE(neva): Fix for compilation with GCC on ARM.
+#if !defined(clang)
+    uint8x16_t lo_nibbles =
+        vmovn_high_u16(vmovn_u16(vreinterpretq_u16_u8(first_batch_lo_nibbles)),
+                       vreinterpretq_u16_u8(second_batch_lo_nibbles));
+#else   // !defined(clang)
     uint8x16_t lo_nibbles = vmovn_high_u16(vmovn_u16(first_batch_lo_nibbles),
                                            second_batch_lo_nibbles);
+#endif  // defined(clang)
 
     // high nibbles are values with odd indexes in loaded batchs.
     // vmovn_high_u16 and vmovn_u16 narrow input words by dropping most
     // significant byte. (0x36, 0x36, 0x36, 0x36, 0x36, 0x37, 0x36, 0x37, 0x36,
     // 0x36, 0x36, 0x36, 0x36, 0x37, 0x36, 0x66)
     uint8x16_t hi_nibbles =
+// NOTE(neva): Fix for compilation with GCC on ARM.
+#if !defined(clang)
+        vmovn_high_u16(vmovn_u16(vreinterpretq_u16_u8(first_batch)),
+                       vreinterpretq_u16_u8(second_batch));
+#else   // !defined(clang)
         vmovn_high_u16(vmovn_u16(first_batch), second_batch);
+#endif  // defined(clang)
 
     // mapping low nibbles to uint8_t values.
     // (0x06, 0x0f, 0x0f, 0x02, 0x01, 0x02, 0x01, 0x02, 0x06, 0x0f, 0x0f, 0x02,
@@ -998,8 +1023,14 @@ bool Uint8ArrayFromHexWithNeon(base::Vector<T>& input_vector,
     // uint8_t values.
     // (0x60, 0x60, 0x60, 0x60, 0x60, 0x70, 0x60, 0x70, 0x60, 0x60, 0x60, 0x60,
     // 0x60, 0x70, 0x60, 0xf0)
+// NOTE(neva): Fix for compilation with GCC on ARM.
+#if !defined(clang)
+    uint8x16_t uint8_shifted_high_nibbles = vreinterpretq_u8_u64(
+        vshlq_n_u64(vreinterpretq_u64_u8(uint8_high_nibbles), 4));
+#else   // !defined(clang)
     uint8x16_t uint8_shifted_high_nibbles =
         vshlq_n_u64(vreinterpretq_u64_u8(uint8_high_nibbles), 4);
+#endif  // defined(clang)
 
     // final result of combining pairs of uint8_t values of low and high
     // nibbles.
-- 
2.34.1

